### PR TITLE
feat: support sparse matrix write (NULL_RATIO) for iotdb-1.3 and iotd…

### PR DESCRIPTION
**Pull request by changxue2022**
…b-2.0

Adds NULL_RATIO config (default 0.0): each cell of every generated row is set to null with probability NULL_RATIO, independently per column, seeded by DATA_SEED for reproducibility.

- core: NULL_RATIO config plumbing, null generation in GenerateDataWorkLoad, and startup guards in ConfigDescriptor.checkConfig (verification modes, non-iotdb-1.3/2.0 switches, and REST / SESSION_BY_RECORD(S) insert modes are rejected with a clear error)
- iotdb-2.0: SessionStrategy.genTablet marks null cells via the measurement name addValue overload (tree and table model); JDBCStrategy emits SQL null literals
- iotdb-1.3: IoTDBSessionBase.genTablet marks null cells in tablet BitMaps (BINARY/DATE slots get a non-null placeholder because the 1.3 session client reads those unconditionally); IoTDB.getInsertOneBatchSql emits SQL null literals
- tests: null generation distribution/determinism, config guards, null serialize round-trip

### CODE REVIEW

File: core/src/main/java/cn/edu/tsinghua/iot/benchmark/workload/GenerateDataWorkLoad.java [hunk 1]
```diff
@@ -79,6 +83,17 @@ protected List<Object> generateOneRow(int deviceIndex, int colIndex, long stepOf
...
       values.add(workloadValues[colIndex][index]);
     }
+    // Sparse matrix write: each cell is null with probability NULL_RATIO, independently. The
+    // values list is a fresh copy, so setting null here never mutates the shared static
+    // workloadValues. When IS_SENSOR_TS_ALIGNMENT is false the list holds a single value and the
+    // coin flip degenerates to a per-value probability.
+    if (config.getNULL_RATIO() > 0) {
+      for (int i = 0; i < values.size(); i++) {
+        if (probTool.returnTrueByProb(config.getNULL_RATIO(), nullRandom)) {
+          values.set(i, null);
```
Comment (SpriCoder) [inline core/src/main/java/cn/edu/tsinghua/iot/benchmark/workload/GenerateDataWorkLoad.java:93]: [P1] Null cells are still counted as successfully written points

This change replaces values with null, but `Batch.pointNum()` and `MultiDeviceBatch.pointNum()` still calculate “number of FIELD columns × number of rows.” `DBWrapper.measureOneBatch()` uses that number for successful point counts and points/s.

For example, a batch with 10 rows and 200 columns contains no actual values when `NULL_RATIO=1`. In a local reproduction, the non-null count was 0, but `pointNum()` returned 2,000. If the write request succeeds, all 2,000 empty positions are therefore counted as successful points. With `NULL_RATIO=0.9`, the reported throughput would typically be about 10 times the non-null point throughput.

Please update both batch implementations to count non-null FIELD values and add tests for fully null and partially null batches. If throughput over all candidate cells is useful, report it as a separate, clearly named metric. This reproduction checks local counting; it does not involve a live database.

**Review by SpriCoder — COMMENTED** [review]
Reviewed commit a36987bc. The inline comments describe three locally reproduced functional issues and one test-coverage suggestion, with concrete examples and proposed fixes.

Validation: all 164 core tests passed on JDK 17. The IoTDB 1.3 module compiled successfully and has no unit tests. IoTDB 2.0 adapter testing could not be completed because dependency downloads stalled. No live database workload was run.

File: core/src/main/java/cn/edu/tsinghua/iot/benchmark/workload/GenerateDataWorkLoad.java [hunk 2]
```diff
@@ -44,6 +44,10 @@ public abstract class GenerateDataWorkLoad extends DataWorkLoad {
   private static final PoissonDistribution poissonDistribution =
       new PoissonDistribution(poissonRandom);
   private static final Random dataRandom = new Random(config.getDATA_SEED());
+  // Random for sparse matrix write (NULL_RATIO), seeded by DATA_SEED so that the null pattern is
+  // deterministic and reproducible. All data clients share the same seed and thus generate the
+  // same null pattern.
+  private final Random nullRandom = new Random(config.getDATA_SEED());
```
Comment (SpriCoder) [inline core/src/main/java/cn/edu/tsinghua/iot/benchmark/workload/GenerateDataWorkLoad.java]: [P2] The same seed does not reproduce null positions with concurrent writers

When `IS_CLIENT_BIND=false`, writer threads share one `SingletonWorkDataWorkLoad`, including this `nullRandom`. The seed fixes the random sequence, but thread scheduling determines which cell receives each random value. The same device, timestamp, and column can therefore be null in one run and populated in another.

I reproduced this by creating a fresh workload for each of two runs with the same configuration and seed, using eight threads. Both runs produced the same 8,000 row identifiers, but 7,949 rows had different null positions. This breaks the reproducibility promised in the PR and can affect comparisons between databases or configurations. The existing determinism test only covers sequential calls.

Please derive the null decision from `DATA_SEED` and the cell coordinates (device, row, and column), rather than the order in which threads consume random numbers. Add a concurrent test that compares null positions for matching row identifiers.

File: core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java [hunk 3]
```diff
@@ -786,6 +792,74 @@ boolean checkConfig() {
...
+    if (nullRatio < 0 || nullRatio > 1) {
+      LOGGER.error(
+          "Invalid parameter NULL_RATIO: {}, whose value range should be [0, 1]", nullRatio);
+      result = false;
+    }
+    BenchmarkMode workMode = config.getBENCHMARK_WORK_MODE();
+    if (workMode == BenchmarkMode.VERIFICATION_WRITE
+        || workMode == BenchmarkMode.VERIFICATION_QUERY) {
+      LOGGER.error(
+          "NULL_RATIO is not supported in {} mode. Please use testWithDefaultPath or generateDataMode.",
```
Comment (SpriCoder) [inline core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java]: [P2] Sparse CSV files can be generated but cannot be read back correctly

This message recommends `generateDataMode`, but the CSV writer and reader do not yet agree on how to represent nulls. `CSVDataWriter` writes the text `null`, while `CSVDataReader` parses it directly as the column type: numeric/date values fail to parse, boolean nulls become false, and text nulls become the string "null".

In a local comparison, a dense batch wrote 10 rows and read back 10 rows. With `NULL_RATIO=1`, it wrote 10 rows but read back 0. Setting `NULL_RATIO` back to 0 before reading passes the configuration guard, but does not fix the nulls already stored in the file.

Please add matching CSV null encoding/decoding and round-trip tests across supported types. If sparse CSV support is outside this PR's scope, reject `generateDataMode` with a nonzero `NULL_RATIO` for now and update this message.

File: core/src/test/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptorTest.java [hunk 4]
```diff
@@ -101,4 +114,117 @@ public void testIoTDBThriftMaxFrameSize() {
...
+   * from {@code checkConfig()} can only come from the NULL_RATIO guards.
+   */
+  private void setUpValidNullRatioContext() {
+    config.setBENCHMARK_WORK_MODE(BenchmarkMode.TEST_WITH_DEFAULT_PATH);
+    config.setIS_CLIENT_BIND(false);
+    config.setIS_DOUBLE_WRITE(false);
+    config.setDEVICE_NUMBER(6000);
+    config.setSCHEMA_CLIENT_NUMBER(1);
+    config.setDATA_CLIENT_NUMBER(1);
+    config.setNULL_RATIO(0.9);
```
Comment (SpriCoder) [inline core/src/test/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptorTest.java:135]: [P3] Add a test that loads NULL_RATIO from the configuration file

The new tests call `setNULL_RATIO(0.9)` directly, bypassing the configuration-file path that users actually use. They could therefore keep passing if the property-loading code were later removed or broken.

Please add a test that writes `NULL_RATIO=0.9` to a temporary configuration file, loads it through the actual configuration entry point, and verifies that the effective value is 0.9. The loading code is currently wired up; this is a coverage suggestion, not a claim that the setting currently fails to load.

File: iotdb-2.0/src/main/java/cn/edu/tsinghua/iot/benchmark/iotdb200/DMLStrategy/SessionStrategy.java [hunk 5]
```diff
@@ -158,63 +158,45 @@ private Tablet genTablet(IBatch batch) {
...
             recordValueIndex++) {
+          Object value = record.getRecordDataValue().get(recordValueIndex);
+          // Sparse matrix write (NULL_RATIO): the measurement-name addValue overload handles
+          // null natively (marks the cell in the tablet BitMaps and stores a typed null
+          // sentinel). A null reaching the typed switch below would NPE, so skip it.
+          if (value == null) {
+            tablet.addValue(sensors.get(sensorIndex).getName(), recordIndex, (Object) null);
+            sensorIndex++;
+            continue;
+          }
```
Comment (jt2594838) [inline iotdb-2.0/src/main/java/cn/edu/tsinghua/iot/benchmark/iotdb200/DMLStrategy/SessionStrategy.java:171]: Not-set values are naturally nulls in a Tablet.
May skip tablet.addValue.

Comment (changxue2022) [reply iotdb-2.0/src/main/java/cn/edu/tsinghua/iot/benchmark/iotdb200/DMLStrategy/SessionStrategy.java:171]: I've just fix them in a commit.

**Review by SpriCoder — COMMENTED** [review]
Good Job!

**Comment by SpriCoder** [comment]
Please take a look at unit test

File: core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java [hunk 6]
```diff
@@ -786,6 +792,75 @@ boolean checkConfig() {
...
+   *       modules handle null values in their write path
+   *   <li>the INSERT_USE_SESSION_TABLET and INSERT_USE_JDBC insert modes
+   *   <li>work modes other than verificationWriteMode / verificationQueryMode
+   * </ul>
+   */
+  private boolean checkNullRatio() {
+    boolean result = true;
+    double nullRatio = config.getNULL_RATIO();
+    // !(nullRatio >= 0 && nullRatio <= 1) rather than (nullRatio < 0 || nullRatio > 1) so that NaN
+    // is rejected too: every comparison with NaN is false, so the latter would let it through.
```
Comment (SpriCoder) [inline core/src/main/java/cn/edu/tsinghua/iot/benchmark/conf/ConfigDescriptor.java]: I think this comment can be removed

**Review by SpriCoder — APPROVED** [review]
Only one thing left
